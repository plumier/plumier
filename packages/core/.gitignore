!src/generic-controller/filter-grammar.js