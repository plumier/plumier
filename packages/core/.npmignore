test
src
.npmignore
.gitignore
tsconfig.json
package-lock.json
tsconfig.build.json
yarn-error.log
yarn.lock