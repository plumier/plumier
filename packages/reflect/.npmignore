.vscode
coverage
scripts
src 
test
.gitignore
.npmignore
.travis.yml
package-lock.json
tsconfig.build.json
tsconfig.json
yarn-error.log
yarn.lock