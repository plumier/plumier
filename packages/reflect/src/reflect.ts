import { ignore, noop, parameterProperties, type } from "./decorators"
import { createClass, metadata, useCache } from "./helpers"
import { parseFunction } from "./parser"
import { Class, ClassReflection, ObjectReflection, Reflection } from "./types"
import { memberVisitors as v, walkTypeMembersRecursive, walkReflectionMembers } from "./walker"

interface TraverseContext {
    path: any[]
}

type Visitor<Ref, Ctx, Ret> = (r: Ref, c: Ctx) => Ret

function pipe<Ref, Ctx, Ret>(visitors: Visitor<Ref, Ctx, Ret>[]): Visitor<Ref, Ctx, Ret> {
    return (value, ctx) => visitors.reduce((a, b) => !!a ? b(a, ctx) : a, value as any)
}

function reflectClass(target: Class): ClassReflection {
    const ref = walkTypeMembersRecursive(target, {
        target, memberVisitor: pipe([
            // add typescript design types information
            v.addsDesignTypes,
            // add metadata decorators
            v.addsDecorators,
            // add datatype info specified by @type decorator
            v.addsTypeOverridden,
            // add parameter properties
            v.addsParameterProperties,
            // add generic type information
            v.addsGenericOverridden,
            // add typeClassification information
            v.addsTypeClassification,
            // remove @ignore decorator
            v.removeIgnored,
        ]),
        classPath: []
    })
    return walkReflectionMembers(ref, {
        classPath: [], target, parent:ref, memberVisitor: pipe([
            v.addsApplyToDecorator,
            // remove generic type overridden, because it may cause issue for addsTypeOverridden
            v.removeGenericTypeOverridden,
            // add datatype info specified by @type decorator
            v.addsTypeOverridden,
            // add typeClassification information
            v.addsTypeClassification,
        ])})
}

function traverseObject(fn: any, name: string, ctx: TraverseContext): Reflection | undefined {
    if (fn === undefined || fn === null) return
    if (Array.isArray(fn)) return
    if (typeof fn === "object") {
        // CIRCULAR: return immediately
        if (ctx.path.some(x => x === fn)) return
        return reflectObject(fn, name, { path: ctx.path.concat(fn) })
    }
    if (typeof fn === "function" && metadata.isConstructor(fn))
        return reflectClass(fn)
    if (typeof fn === "function")
        return parseFunction(fn)
}

function reflectObject(object: any, name: string, ctx: TraverseContext): ObjectReflection {
    return {
        kind: "Object", name,
        members: Object.keys(object).map(x => traverseObject(object[x], x, ctx)).filter((x): x is Reflection => !!x)
    }
}

function reflectModuleOrClass(opt: string | Class) {
    if (typeof opt === "string") {
        return reflectObject(require(opt), "module", { path: [] })
    }
    else {
        return reflectClass(opt)
    }
}


// --------------------------------------------------------------------- //
// ------------------------------- CACHE ------------------------------- //
// --------------------------------------------------------------------- //

interface ReflectOption {
    /**
     * Flush cached current module/class metadata before processing. 
     */
    flushCache?: true
}

const cacheStore = new Map<string | Class, ClassReflection | ObjectReflection>()
const reflectCached = useCache(cacheStore, reflectModuleOrClass, x => x)

/**
 * Reflect module
 * @param path module name
 */
function reflect(path: string, opt?: Partial<ReflectOption>): ObjectReflection

/**
 * Reflect class
 * @param classType Class 
 */
function reflect(classType: Class, opt?: Partial<ReflectOption>): ClassReflection

function reflect(pathOrClass: string | Class, opt?: Partial<ReflectOption>): ClassReflection | ObjectReflection {
    if (opt?.flushCache)
        cacheStore.delete(pathOrClass)
    return reflectCached(pathOrClass)
}

function flush(pathOrClass?: string | Class) {
    if (pathOrClass)
        cacheStore.delete(pathOrClass)
    else
        cacheStore.clear()
}

// --------------------------------------------------------------------- //
// ----------------------------- DECORATORS ---------------------------- //
// --------------------------------------------------------------------- //

/**
 * Decorator that do nothing, intended to be able to inspect data type
 */
reflect.noop = noop

/**
 * Ignore member from metadata generated
 */
reflect.ignore = ignore

/**
 * Override type definition information. Useful to add type definition for some data type that is erased 
 * after transfile such as Partial<Type> or ReadOnly<Type>
 * 
 * If applied to parameter it will override the parameter type
 * 
 * If applied to property it will override the property type
 * 
 * if applied to method it will overrid the method return value
 * @param type The type overridden
 * @param info Additional information about type (readonly, partial etc)
 */
reflect.type = type

/**
 * Create class dynamically
 */
reflect.create = createClass

/**
 * Mark all constructor parameters as properties
 */
reflect.parameterProperties = parameterProperties

/**
 * Flush metadata cache, call without parameter to clear all
 */
reflect.flush = flush

export { reflect }
