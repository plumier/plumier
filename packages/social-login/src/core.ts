import {
    ActionResult,
    bind,
    CustomMiddleware,
    DefaultFacility,
    HttpCookie,
    HttpStatusError,
    Invocation,
    invoke,
    PlumierApplication,
    response,
    RouteInfo,
    RouteMetadata,
    VirtualRoute,
    Class
} from "@plumier/core"
import Axios, { AxiosError } from "axios"
import crypto from "crypto"
import Csrf from "csrf"
import debug from "debug"
import { Context } from "koa"
import OAuth from "oauth-1.0a"
import qs from "querystring"
import { decorateMethod } from "tinspector"

// --------------------------------------------------------------------- //
// ------------------------------- TYPES ------------------------------- //
// --------------------------------------------------------------------- //

export enum CookieName {
    csrfSecret = "plum-oauth:csrf-secret",
    provider = "plum-oauth:provider"
}

export type SocialProvider = "Facebook" | "GitHub" | "Google" | "GitLab" | "Twitter"
export type OAuthVersion = "1.0a" | "2.0"
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type Optional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

export interface EndPoint {
    endpoint: string
    params: {}
}

export interface OAuthOptions {
    provider: SocialProvider
    clientId: string
    clientSecret: string
    token: EndPoint
    requestToken?: EndPoint
    profile: EndPoint & { transformer: (value: any) => Optional<OAuthUser, "provider"> }
    login: EndPoint
    oAuthVersion: OAuthVersion
}

export interface OAuthUser {
    provider: SocialProvider
    id: string,
    name: string,
    firstName: string,
    lastName: string,
    profilePicture: string,
    email?: string,
    gender?: string,
    dateOfBirth?: string
}

export interface OAuthProviderOption {
    clientId?: string,
    clientSecret?: string,
    loginEndPoint?: string,
    profileParams?: {}
}

interface ParserResult {
    user: Optional<OAuthUser, "provider">,
    profile: any,
    token: any
}

interface GeneratorResult {
    redirect: string,
    cookies: OAuthCookies
}

export type OAuth1Request = (opt: { url: string, method: "GET" | "POST", data: any }, token?: OAuth.Token) => Promise<any>

// --------------------------------------------------------------------- //
// ------------------------------- HELPER ------------------------------ //
// --------------------------------------------------------------------- //

const log = {
    debug: debug("@plumier/social-login:debug"),
    error: debug("@plumier/social-login:error")
}

export function splitName(name: string | undefined) {
    if (!name) return { firstName: "", lastName: "" }
    const idx = name.lastIndexOf(" ")
    if (idx === -1) return { firstName: name, lastName: "" }
    const lastName = name.substr(idx + 1)
    const firstName = name.substring(0, idx)
    return { firstName, lastName }
}

class OAuthCookies {
    static async oAuth2(provider: SocialProvider) {
        const csrf = new Csrf()
        const secret = await csrf.secret()
        return new OAuthCookies(secret, provider)
    }

    constructor(public secret: string, private provider: SocialProvider) { }

    toHttpCookies(): HttpCookie[] {
        return [
            { key: CookieName.csrfSecret, value: this.secret },
            { key: CookieName.provider, value: this.provider },
        ]
    }
}

export function oAuth1ARequest(apiKey: string, apiKeySecret: string): OAuth1Request {
    var oauth = new OAuth({
        consumer: {
            key: apiKey,
            secret: apiKeySecret
        },
        signature_method: 'HMAC-SHA1',
        hash_function: function (base_string, key) {
            return crypto.createHmac('sha1', key).update(base_string).digest('base64');
        }
    });
    return async ({ url, method, data }: { url: string, method: "GET" | "POST", data: any }, token?: OAuth.Token) => {
        const headers = oauth.toHeader(oauth.authorize({ url, method, data }, token))
        const resp = await Axios({ url, method, headers, data: method === "GET" ? undefined : data })
        return resp.data
    }
}

// --------------------------------------------------------------------- //
// ----------------------------- DECORATORS ---------------------------- //
// --------------------------------------------------------------------- //

export interface OAuthRedirectUriDecorator { name: "OAuthRedirectUriDecorator", provider: SocialProvider }

export function redirectUri(provider: SocialProvider | "General" = "General") {
    return decorateMethod(<OAuthRedirectUriDecorator>{ name: "OAuthRedirectUriDecorator", provider })
}

declare module "@plumier/core" {
    namespace bind {
        export function oAuthUser(): (target: any, name: string, index: number) => void
        export function oAuthToken(): (target: any, name: string, index: number) => void
        export function oAuthProfile(): (target: any, name: string, index: number) => void
    }
}

bind.oAuthUser = () => bind.custom(x => x.state.oAuthUser)
bind.oAuthToken = () => bind.custom(x => x.state.oAuthToken)
bind.oAuthProfile = () => bind.custom(x => x.state.oAuthProfile)



// --------------------------------------------------------------------- //
// ------------------------------ RESPONSE ----------------------------- //
// --------------------------------------------------------------------- // 

declare module "@plumier/core" {
    namespace response {
        function postMessage(message: any, origin?: string): ActionResult;
    }
}

response.postMessage = (message: any, origin?: string) => {
    return new ActionResult(`
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class="container"></div>
            <script type="text/javascript">
                var message = '${JSON.stringify(message)}';
                var origin ${!!origin ? `= '${origin}'` : ""};
                (function(){
                    window.onbeforeunload = function () {
                        window.opener.postMessage({ status: "Close" }, origin || window.location.origin);
                    };
                    window.opener.postMessage(JSON.parse(message), origin || window.location.origin);
                })()
            </script>
        </body>
    </html>
    `).setHeader("Content-Type", "text/html")
}

function clientRedirect(url: string) {
    // redirect client using JS to possibly set cookie before redirection
    return new ActionResult(`
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class="container"></div>
            <script type="text/javascript">
                const REDIRECT = '${url}';
                (function(){
                    window.location.href = REDIRECT;
                })()
            </script>
        </body>
    </html>
    `).setHeader("Content-Type", "text/html")
}

// --------------------------------------------------------------------- //
// ------------------------ LOGIN URL GENERATOR ------------------------ //
// --------------------------------------------------------------------- //

// oauth 2.0
async function oAuth20LoginGenerator(ctx: Context, option: OAuthOptions, redirect_uri: string): Promise<GeneratorResult> {
    const params: any = { ...option.login.params, ...ctx.query }
    params.redirect_uri = ctx.origin + redirect_uri
    params.client_id = option.clientId
    const cookies = await OAuthCookies.oAuth2(option.provider)
    const csrf = new Csrf()
    params.state = csrf.create(cookies.secret)
    const redirect = option.login.endpoint + "?" + qs.stringify(params)
    return { redirect, cookies }
}

// oauth 1.0a
async function oAuth10aLoginGenerator(ctx: Context, opt: OAuthOptions, redirectUri: string): Promise<GeneratorResult> {
    const request = oAuth1ARequest(opt.clientId, opt.clientSecret)
    const rawTokens = await request({
        url: opt.requestToken!.endpoint,
        method: "POST",
        data: { oauth_callback: ctx.origin + redirectUri }
    })
    const reqTokens: any = qs.parse(rawTokens)
    const params = { oauth_token: reqTokens.oauth_token, ...opt.login.params, ...ctx.query }
    const redirect = `${opt.login.endpoint}?${qs.stringify(params)}`
    const cookies = new OAuthCookies(reqTokens.oauth_token, opt.provider)
    return { redirect, cookies }
}


// --------------------------------------------------------------------- //
// --------------------------- PROFILE PARSER -------------------------- //
// --------------------------------------------------------------------- //

// oauth 2.0
async function oAuth20ProfileParser(ctx: Context, option: OAuthOptions, redirect_uri: string): Promise<ParserResult> {
    const req = ctx.request
    const secret = ctx.cookies.get(CookieName.csrfSecret)!
    if (!new Csrf().verify(secret, req.query.state)) {
        log.error("*** Invalid csrf token ***")
        log.error("CSRF Secret: %s", secret)
        log.error("CSRF Token: %s", req.query.state)
        throw new HttpStatusError(400)
    }
    const { clientId: client_id, clientSecret: client_secret } = option
    // request access token
    const { data: token } = await Axios.post<{ access_token: string }>(option.token.endpoint,
        { client_id, redirect_uri, client_secret, code: req.query.code, ...option.token.params },
        { headers: { Accept: 'application/json' } })
    // request profile
    const { data: profile } = await Axios.get(option.profile.endpoint, {
        params: option.profile.params,
        headers: { Authorization: `Bearer ${token.access_token}` }
    })
    const user = option.profile.transformer(profile)
    return { user, profile, token }
}

// oauth 1.0
async function oAuth10aProfileParser(ctx: Context, option: OAuthOptions, redirect_uri: string): Promise<ParserResult> {
    const req = ctx.request
    const secret = ctx.cookies.get(CookieName.csrfSecret)!
    if (secret !== req.query.oauth_token) {
        log.error("*** Invalid oauth_token ***")
        log.error("Saved token: %s", secret)
        log.error("Request token: %s", req.query.oauth_token)
        throw new HttpStatusError(400)
    }
    const request = oAuth1ARequest(option.clientId, option.clientSecret)
    const rawTokens = await request({
        url: option.token.endpoint, method: "POST",
        data: {
            oauth_token: req.query.oauth_token,
            oauth_verifier: req.query.oauth_verifier
        }
    })
    const token: any = qs.parse(rawTokens)
    const profile = await request({
        url: option.profile.endpoint, method: "GET",
        data: {}
    }, { key: token.oauth_token, secret: token.oauth_token_secret })
    const user = option.profile.transformer(profile)
    return { user, profile, token }
}

// --------------------------------------------------------------------- //
// ---------------------------- MIDDLEWARES ---------------------------- //
// --------------------------------------------------------------------- //

class OAuthLoginEndPointMiddleware implements CustomMiddleware {
    readonly generator = {
        "1.0a": oAuth10aLoginGenerator,
        "2.0": oAuth20LoginGenerator
    }

    constructor(private option: OAuthOptions, private loginPath: string, private redirectUri: string) { }

    async execute(invocation: Readonly<Invocation>): Promise<ActionResult> {
        if (invocation.ctx.path.toLowerCase() !== this.loginPath.toLowerCase()) return invocation.proceed()
        const { redirect, cookies } = await this.generator[this.option.oAuthVersion](invocation.ctx,
            this.option, this.redirectUri)
        return clientRedirect(redirect)
            .setCookie(cookies.toHttpCookies())
    }
}

class OAuthRedirectUriMiddleware implements CustomMiddleware {
    readonly parser = {
        "1.0a": oAuth10aProfileParser,
        "2.0": oAuth20ProfileParser
    }

    constructor(private option: OAuthOptions, private redirectUri: string) { }

    async execute(inv: Readonly<Invocation>): Promise<ActionResult> {
        const req = inv.ctx.request;
        if (inv.ctx.state.caller === "invoke") return inv.proceed()
        if (inv.ctx.path.toLocaleLowerCase() !== this.redirectUri.toLowerCase()) return inv.proceed()
        const provider = inv.ctx.cookies.get(CookieName.provider)
        if (provider !== this.option.provider) return inv.proceed()
        log.debug("Provider: %s", this.option.provider)
        const redirectUri = req.origin + req.path
        try {
            const result = await this.parser[this.option.oAuthVersion](inv.ctx, this.option, redirectUri)
            log.debug("OAuth User: %O", result.user)
            const provider = this.option.provider
            inv.ctx.state.oAuthUser = { ...result.user, provider }
            inv.ctx.state.oAuthToken = { ...result.token, provider }
            inv.ctx.state.oAuthProfile = { ...result.profile, provider }
        }
        catch (e) {
            log.error("*** OAuth Error ***", this.option.provider)
            log.error("Auth Code: %s", req.query.code)
            log.error("Redirect Uri: %s", redirectUri)
            if (e.response) {
                const error = (e as AxiosError)
                const response = error.response!;
                log.error("Request URL: %s %s", error.config.method, error.config.url)
                log.error("Request Data: %o", error.config.data)
                log.error("Request Headers: %o", error.config.headers)
                log.error("Error: %o", { status: response.status, message: response.statusText, data: response.data })
                throw new HttpStatusError(response.status, response.statusText)
            }
            throw e
        }
        return invoke(inv.ctx, inv.ctx.route!)
    }
}

// --------------------------------------------------------------------- //
// ------------------------------ FACILITY ----------------------------- //
// --------------------------------------------------------------------- //

export class OAuthProviderBaseFacility extends DefaultFacility {
    private option: Optional<OAuthOptions, "clientId" | "clientSecret">
    private loginEndpoint: string

    constructor(option: Optional<OAuthOptions, "clientId" | "clientSecret">, opt?: OAuthProviderOption) {
        super()
        this.option = {
            ...option,
            clientId: opt?.clientId,
            clientSecret: opt?.clientSecret,
        }
        if (opt && opt.profileParams)
            this.option.profile.params = { ...this.option.profile.params, ...opt.profileParams }
        this.loginEndpoint = opt?.loginEndPoint ?? `/auth/${this.option.provider.toLowerCase()}/login`
    }

    async generateRoutes(): Promise<VirtualRoute[]> {
        return [{
            kind: "VirtualRoute",
            method: "get",
            provider: this.constructor as Class,
            url: this.loginEndpoint,
            access: "Public",
            overridable: false,
            openApiOperation: {
                description: `Redirect request to ${this.option.provider} login page. This endpoint writes CSRF token on the client required to generate ${this.option.provider} login endpoint`,
                tags: ["Social Login"],
                parameters: [
                    {}
                ],
            }
        }]
    }

    getRedirectUri(routes: RouteMetadata[], provider: string) {
        return routes.find((x): x is RouteInfo => x.kind === "ActionRoute" && x.action.decorators
            .some((y: OAuthRedirectUriDecorator) => y.name === "OAuthRedirectUriDecorator" && y.provider === provider))
    }

    async initialize(app: Readonly<PlumierApplication>, routes: RouteMetadata[]) {
        const redirectUriRoute = this.getRedirectUri(routes, this.option.provider) ?? this.getRedirectUri(routes, "General")
        if (!redirectUriRoute) throw new Error(`OAuth: No ${this.option.provider} redirect uri handler found`)
        if (redirectUriRoute.url.search(":") > -1) throw new Error(`OAuth: Parameterized route is not supported on ${this.option.provider} callback uri`)
        if (redirectUriRoute.method !== "get") throw new Error(`OAuth: Redirect uri must have GET http method on ${redirectUriRoute.controller.name}.${redirectUriRoute.action.name}`)
        const clientIdKey = `PLUM_${this.option.provider.toUpperCase()}_CLIENT_ID`
        const clientSecretKey = `PLUM_${this.option.provider.toUpperCase()}_CLIENT_SECRET`
        const clientId = this.option.clientId ?? process.env[clientIdKey]
        if (!clientId) throw new Error(`OAuth: Client id for ${this.option.provider} not provided`)
        const clientSecret = this.option.clientSecret ?? process.env[clientSecretKey]
        if (!clientSecret) throw new Error(`OAuth: Client secret for ${this.option.provider} not provided`)
        const option = { ...this.option, clientId, clientSecret }
        app.use(new OAuthLoginEndPointMiddleware(option, this.loginEndpoint, redirectUriRoute.url))
        app.use(new OAuthRedirectUriMiddleware(option, redirectUriRoute.url))
    }
}