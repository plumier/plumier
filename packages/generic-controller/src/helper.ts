import { RelationDecorator, RelationPropertyDecorator } from "@plumier/core"
import reflect, { Class, generic, GenericTypeDecorator } from "@plumier/reflect"


// --------------------------------------------------------------------- //
// ----------------------- ENTITY RELATION HELPER ---------------------- //
// --------------------------------------------------------------------- //

const genericControllerRegistry = new Map<Class, boolean>()

function updateGenericControllerRegistry(cls: Class) {
    genericControllerRegistry.set(cls, true)
}

function getGenericControllerRelation(controller: Class) {
    const meta = reflect(controller)
    const types = generic.getGenericTypeParameters(controller)
    const parentEntityType = types[0]
    const entityType = types[1]
    const oneToMany = meta.decorators.find((x: RelationPropertyDecorator): x is RelationPropertyDecorator => x.kind === "plumier-meta:relation-prop-name")
    const relation = oneToMany!.name
    return { parentEntityType, entityType, relation, inverseProperty: oneToMany!.inverseProperty }
}

function getGenericControllerInverseProperty(controller: Class) {
    const rel = getGenericControllerRelation(controller)
    return rel.inverseProperty
}

function getGenericControllerOneToOneRelations(type: Class) {
    const meta = reflect(type)
    const result = []
    for (const prop of meta.properties) {
        if (prop.decorators.find((x: RelationDecorator) => x.kind === "plumier-meta:relation") && prop.typeClassification !== "Array") {
            result.push(prop)
        }
    }
    return result
}


export {
    getGenericControllerOneToOneRelations, getGenericControllerInverseProperty,
    getGenericControllerRelation, updateGenericControllerRegistry, genericControllerRegistry
}