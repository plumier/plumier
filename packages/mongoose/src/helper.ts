import { authorize } from "@plumier/core"
import reflect, { Class, useCache } from "@plumier/reflect"

import { RefDecorator } from "./types"

function normalizeEntityNoCache(type: Class) {
    const meta = reflect(type)
    reflect.flush(type)
    for (const prop of meta.properties) {
        const ref = prop.decorators.find((x: RefDecorator) => x.name === "MongooseRef")
        if (ref && prop.typeClassification === "Array") {
            const decorators = [
                authorize.readonly(),
                authorize.writeonly()
            ]
            Reflect.decorate(decorators, type.prototype, prop.name, void 0)
        }
    }
    return { success: true }
}

const normalizeEntityCache = new Map<Class, any>()

export const normalizeEntity = useCache(normalizeEntityCache, normalizeEntityNoCache, x => x)