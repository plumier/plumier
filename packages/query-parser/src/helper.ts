import { AuthorizeDecorator, AuthorizerContext, authPolicy, AuthPolicy, EntityAuthPolicy, executeAuthorizer } from "@plumier/core"
import { Class, generic } from "@plumier/reflect"


const ParserAst = Symbol()

function getDecoratorType(controller: Class, expType: string | Class) {
    // extract generic type from controller if string type provided
    return typeof expType === "string" ? generic.getGenericType(controller, expType) as Class : expType
}

async function isAuthorized(decorators: AuthorizeDecorator[], ctx: AuthorizerContext): Promise<boolean> {
    const authPolicies = ctx.ctx.config.authPolicies.map(x => new x())
    const policies: { decorator: AuthorizeDecorator, isEntityPolicy: boolean }[] = []
    const staticPolicies = []
    // flatten decorator to make each decorator only have one policy
    // and add flag if its an entity policy
    // to make them easily filtered by its flag
    for (const decorator of decorators) {
        for (const policy of decorator.policies) {
            const isEntityPolicy = authPolicies.some(x => x.name === policy && x instanceof EntityAuthPolicy)
            const dec: AuthorizeDecorator = { ...decorator, policies: [policy] }
            if (!isEntityPolicy)
                staticPolicies.push(dec)
            policies.push({ decorator: dec, isEntityPolicy })
        }
    }
    if (policies.every(x => x.isEntityPolicy)) return false
    return executeAuthorizer(staticPolicies, ctx)
}

export { getDecoratorType, ParserAst, isAuthorized }