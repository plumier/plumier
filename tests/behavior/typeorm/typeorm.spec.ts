import { Entity, Column, PrimaryGeneratedColumn, getConnection, OneToMany, OneToOne, JoinColumn, ManyToMany, ManyToOne, JoinTable } from "typeorm"
import { fixture } from '../helper'
import { TypeORMFacility, CRUDTypeORMFacility, } from '@plumier/typeorm'
import reflect from "tinspector"
import { Class, consoleLog } from '@plumier/core'
import Plumier, { WebApiFacility } from '@plumier/plumier'

describe("TypeOrm", () => {
    function createApp(entities: Function[]) {
        class UsersController {
            get() { }
        }
        return fixture(UsersController)
            .set(new TypeORMFacility({
                type: "sqlite",
                database: ":memory:",
                dropSchema: true,
                entities: entities,
                synchronize: true,
                logging: false
            }))
            .initialize()
    }
    afterEach(async () => {
        let conn = getConnection();
        if (conn.isConnected)
            await conn.close();
    });
    describe("Facility", () => {
        function extract(type: Class) {
            return reflect(type).properties.map(({ name, type }) => ({ name, type }))
        }
        it("Should able to reflect entity properties", async () => {
            @Entity()
            class MyEntity {
                @PrimaryGeneratedColumn()
                id: number = 123
                @Column()
                num: number = 200
            }
            await createApp([MyEntity])
            expect(extract(MyEntity)).toMatchSnapshot()
        })
        it("Should able to reflect entity properties with inheritance", async () => {
            @Entity()
            class MyEntity {
                @PrimaryGeneratedColumn()
                id: number = 123
            }
            @Entity()
            class ChildEntity extends MyEntity {
                @Column()
                num: number = 200
            }
            await createApp([ChildEntity, MyEntity])
            expect(extract(ChildEntity)).toMatchSnapshot()
        })
        it("Should able to reflect one to one relation", async () => {
            @Entity()
            class Parent {
                @PrimaryGeneratedColumn()
                id: number = 123
                @Column()
                name: string = "mimi"
            }
            @Entity()
            class MyEntity {
                @PrimaryGeneratedColumn()
                id: number = 123
                @OneToOne(x => Parent)
                @JoinColumn()
                parent: Parent = {} as any
            }
            await createApp([MyEntity, Parent])
            expect(extract(MyEntity)).toMatchSnapshot()
        })
        it("Should able to reflect one to one relation with inverse relation", async () => {
            @Entity()
            class Parent {
                @PrimaryGeneratedColumn()
                id: number = 123
                @Column()
                name: string = "mimi"
            }
            @Entity()
            class MyEntity {
                @PrimaryGeneratedColumn()
                id: number = 123
                @OneToOne(x => Parent)
                @JoinColumn()
                parent: Parent
            }
            await createApp([Parent, MyEntity])
            expect(extract(MyEntity)).toMatchSnapshot()
        })
        it("Should able to reflect one to many relation", async () => {
            @Entity()
            class MyEntity {
                @PrimaryGeneratedColumn()
                id: number = 123
                @OneToMany(x => Child, x => x.entity)
                parent: Child[]
            }
            @Entity()
            class Child {
                @PrimaryGeneratedColumn()
                id: number
                @Column()
                name: string
                @ManyToOne(x => MyEntity, x => x.parent)
                entity: MyEntity
            }
            await createApp([MyEntity, Child])
            expect(extract(MyEntity)).toMatchSnapshot()
            expect(extract(Child)).toMatchSnapshot()
        })
        it("Should able to reflect many to many relation", async () => {
            @Entity()
            class MyEntity {
                @PrimaryGeneratedColumn()
                id: number = 123
                @ManyToMany(x => Child)
                @JoinTable()
                parent: Child[]
            }
            @Entity()
            class Child {
                @PrimaryGeneratedColumn()
                id: number
                @Column()
                name: string
            }
            await createApp([MyEntity, Child])
            expect(extract(MyEntity)).toMatchSnapshot()
        })
        it("Should throw error when no option specified", async () => {
            const fn = jest.fn()
            class UsersController {
                get() { }
            }
            try {
                await fixture(UsersController)
                    .set(new TypeORMFacility())
                    .initialize()
            }
            catch (e) {
                fn(e)
            }
            expect(fn.mock.calls).toMatchSnapshot()
        })
    })
    describe("CRUD", () => {
        function createApp(entities: Function[]) {
            return new Plumier()
                .set(new WebApiFacility())
                .set(new CRUDTypeORMFacility({
                    type: "sqlite",
                    database: ":memory:",
                    dropSchema: true,
                    entities: entities,
                    synchronize: true,
                    logging: false
                }))
                .initialize()
        }
        it.only("Should generate routes properly", async () => {
            @Entity()
            class User {
                @PrimaryGeneratedColumn()
                id:number 
                @Column()
                email:string 
                @Column()
                name:string
            }
            const mock = consoleLog.startMock()
            await createApp([User])
            expect(mock.mock.calls).toMatchSnapshot()
            consoleLog.clearMock()
        })
    })
})


