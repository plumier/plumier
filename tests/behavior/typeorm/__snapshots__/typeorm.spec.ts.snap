// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeOrm CRUD Route Generator Should generate routes from inheritance 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. DynamicType.list(offset, limit) -> GET    /users",
  ],
  Array [
    "2. DynamicType.save(data)          -> POST   /users",
  ],
  Array [
    "3. DynamicType.get(id)             -> GET    /users/:id",
  ],
  Array [
    "4. DynamicType.modify(id, data)    -> PUT    /users/:id",
  ],
  Array [
    "5. DynamicType.modify(id, data)    -> PATCH  /users/:id",
  ],
  Array [
    "6. DynamicType.delete(id)          -> DELETE /users/:id",
  ],
  Array [],
]
`;

exports[`TypeOrm CRUD Route Generator Should generate routes from multiple entities 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    " 1. DynamicType.list(offset, limit) -> GET    /users",
  ],
  Array [
    " 2. DynamicType.save(data)          -> POST   /users",
  ],
  Array [
    " 3. DynamicType.get(id)             -> GET    /users/:id",
  ],
  Array [
    " 4. DynamicType.modify(id, data)    -> PUT    /users/:id",
  ],
  Array [
    " 5. DynamicType.modify(id, data)    -> PATCH  /users/:id",
  ],
  Array [
    " 6. DynamicType.delete(id)          -> DELETE /users/:id",
  ],
  Array [
    " 7. DynamicType.list(offset, limit) -> GET    /secondusers",
  ],
  Array [
    " 8. DynamicType.save(data)          -> POST   /secondusers",
  ],
  Array [
    " 9. DynamicType.get(id)             -> GET    /secondusers/:id",
  ],
  Array [
    "10. DynamicType.modify(id, data)    -> PUT    /secondusers/:id",
  ],
  Array [
    "11. DynamicType.modify(id, data)    -> PATCH  /secondusers/:id",
  ],
  Array [
    "12. DynamicType.delete(id)          -> DELETE /secondusers/:id",
  ],
  Array [],
]
`;

exports[`TypeOrm CRUD Route Generator Should generate routes properly 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. DynamicType.list(offset, limit) -> GET    /users",
  ],
  Array [
    "2. DynamicType.save(data)          -> POST   /users",
  ],
  Array [
    "3. DynamicType.get(id)             -> GET    /users/:id",
  ],
  Array [
    "4. DynamicType.modify(id, data)    -> PUT    /users/:id",
  ],
  Array [
    "5. DynamicType.modify(id, data)    -> PATCH  /users/:id",
  ],
  Array [
    "6. DynamicType.delete(id)          -> DELETE /users/:id",
  ],
  Array [],
]
`;

exports[`TypeOrm Facility Should able to reflect entity properties 1`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
  Object {
    "name": "num",
    "type": Number,
  },
]
`;

exports[`TypeOrm Facility Should able to reflect entity properties with inheritance 1`] = `
Array [
  Object {
    "name": "num",
    "type": Number,
  },
  Object {
    "name": "id",
    "type": Number,
  },
]
`;

exports[`TypeOrm Facility Should able to reflect many to many relation 1`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
  Object {
    "name": "parent",
    "type": Array [
      Child,
    ],
  },
]
`;

exports[`TypeOrm Facility Should able to reflect one to many relation 1`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
  Object {
    "name": "parent",
    "type": Array [
      Child,
    ],
  },
]
`;

exports[`TypeOrm Facility Should able to reflect one to many relation 2`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
  Object {
    "name": "name",
    "type": String,
  },
  Object {
    "name": "entity",
    "type": MyEntity,
  },
]
`;

exports[`TypeOrm Facility Should able to reflect one to one relation 1`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
  Object {
    "name": "parent",
    "type": Parent,
  },
]
`;

exports[`TypeOrm Facility Should able to reflect one to one relation with inverse relation 1`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
  Object {
    "name": "parent",
    "type": Parent,
  },
]
`;

exports[`TypeOrm Facility Should throw error when no option specified 1`] = `
Array [
  Array [
    [Error: No connection options were found in any orm configuration files.],
  ],
]
`;
